custom_random_forest.py

```diff
@@ -0,0 +1,66 @@
+from sklearn.tree import DecisionTreeClassifier
+import numpy as np
+from sklearn.base import BaseEstimator
+from multiprocessing import Pool
+
+
+class RandomForestClassifierCustom(BaseEstimator):
+    def __init__(self, n_estimators=10, max_depth=None, max_features=None, random_state=None):
+        self.n_estimators = n_estimators
+        self.max_depth = max_depth
+        self.max_features = max_features
+        self.random_state = random_state
+
+        self.trees = []
+        self.feat_ids_by_tree = []
+
+    def fit(self, X, y, n_jobs=1):
+        n_features = X.shape[1]
+        self.classes_ = np.unique(y)
+        self.trees = []
+        self.feat_ids_by_tree = []
+
+        with Pool(n_jobs) as pool:
+            output = pool.map(self.fit_one_tree, [(X, y, i) for i in range(self.n_estimators)])
+
+        self.trees, self.feat_ids_by_tree = zip(*output)
+        return self
+
+    def fit_one_tree(self, args):
+        X, y, i = args
+        n_features = X.shape[1]
+
+        np.random.seed(self.random_state + i)
+        if self.max_features is None:
+            self.max_features = n_features
+        feat_ids = np.random.choice(n_features, self.max_features, replace=False)
+        bootstrap_indices = np.random.choice(np.arange(X.shape[0]), size=X.shape[0], replace=True)
+        X_bootstrap = X[bootstrap_indices][:, feat_ids]
+        y_bootstrap = y[bootstrap_indices]
+        tree = DecisionTreeClassifier(max_depth=self.max_depth, random_state=self.random_state)
+        tree.fit(X_bootstrap, y_bootstrap)
+        return tree, feat_ids
```

Comment:
Супер. Только я бы еще по смыслу добавил `_` в названии метода чтобы подчеркнуть что это не часть публичного интерфейса, а еще переменная `i` не очень хорошее имя. Это же не просто счетчик. Можно, нарпимер, `tree_id`.

```suggestion
    def _fit_one_tree(self, args):
        X, y, tree_id = args
        n_features = X.shape[1]

        np.random.seed(self.random_state + tree_id)
        if self.max_features is None:
            self.max_features = n_features
        feat_ids = np.random.choice(n_features, self.max_features, replace=False)
        bootstrap_indices = np.random.choice(np.arange(X.shape[0]), size=X.shape[0], replace=True)
        X_bootstrap = X[bootstrap_indices][:, feat_ids]
        y_bootstrap = y[bootstrap_indices]
        tree = DecisionTreeClassifier(max_depth=self.max_depth, random_state=self.random_state)
        tree.fit(X_bootstrap, y_bootstrap)
        return tree, feat_ids
```